functions/dialogflow/functions/index.js

```diff
@@ -0,0 +1,161 @@
+      res.on('end', () => {
+        // Next step in pipeline
+        saveOAuthToken(context, JSON.parse(oauthToken));
+        return resolve(JSON.parse(oauthToken));
+      })
+    });
+ 
+    post_req.on('error', (e) => {
+      console.log('ERROR generating new token', e.message);
+      return "Error retrieving token";
+    });
+ 
+    // Sets up the scope that we want the end user to have.
+    const body = `{ 
+      "delegates": [],
+      "scope": [
+        "https://www.googleapis.com/auth/dialogflow"
+      ],
+        "lifetime": "3599s"
+    }`;
+ 
+    // post the data
+    post_req.write(body);
+    post_req.end();
+  });
+}
+ 
+ 
+function retrieveCredentials(context) {
+  return new Promise((resolve, reject) => {
+    // To create a new access token, we first have to retrieve the credentials
+    // of the service account that will make the generateTokenRequest(). 
+    // To do that, we will use the App Engine Default Service Account.
+    var options = {
+      host: 'metadata.google.internal',
+      path: '/computeMetadata/v1/instance/service-accounts/default/token',
+      method: 'GET',
+      headers: { 'Metadata-Flavor': 'Google' }
+    };
+ 
+    var get_req = http.get(options, (res) => {
+      var body = '';
+ 
+      res.on('data', (chunk) => {
+        body += chunk;
+      });
+ 
+      res.on('end', () => {
+        const response = JSON.parse(body);
+        return generateAccessToken(context, response.access_token, response.token_type).then((result) => {
+          return resolve(result);
+        });
+      });
+    });
+    get_req.on('error', (e) => {
+      console.log("Error retrieving credentials", e.message);
+      return "Error retrieving token";
+    });
+    get_req.end();
+  });
+}
+ 
+ 
+// This method verifies the token expiry by validating against current time
+function isValid(expiryTime) {
+  var currentDate = new Date();
+  var expirationDate = new Date(expiryTime);
+  // If within 5 minutes of expiration, return false
+  return currentDate <= (expirationDate - 1000 * 60 * 5);
+}
+ 
+ 
+exports.getOAuthToken = functions.https.onCall((data, context) => {
+  // Checking that the user is authenticated.
+  if (!context.auth) {
+    // Throwing an HttpsError so that the client gets the error details.
+    throw new functions.https.HttpsError('failed-precondition', 'The function must be called ' +
+      'while authenticated.');
+  }
+  // Retrieve the token from the database
+  var docRef = db.collection('DialogflowTokens').doc("OauthToken");
+
+  return docRef.get().then((doc) => {
+    if (doc.exists === false) {
+      return retrieveCredentials(context).then((result) => {
+        console.log("Received a new token when there was none", result);
+        return result;
+      });
+    }
+    // Check if the token is expired 
+    if (isValid(doc.data().token.expireTime)) {
+      return doc.data().token;
+    } else { // Generate a new token
+      // eslint-disable-next-line promise/no-nesting
```

Comment:
Not sure if this line is still needed with the top level comment at the start of the file.